library/lists.pl

```diff
@@ -552,6 +553,21 @@
     Sum1 is Sum0 + X,
     sum_list(Xs, Sum1, Sum).
 
+%!  prod_list(+List, -Sum) is det.
+%
+%   Prod is the result of multiplying all numbers in List.
+
+prod_list(Xs, Prod) :-
+    prod_list(Xs, 1, Prod).
+
+prod_list([], Prod, Prod).
+prod_list([X|Xs], Prod0, Prod) :-
+    (   X =\= 0 ->
```

Comment:
This optimization should be used only if it can be justified by benchmarks. My guess is that it isn't justified.

Reply:
In the latest commit there is indeed an improvement. With the 0 check (current version)
```
test:-
    length(L,1000),
    maplist(random(0,3),L), !,
    time(prod_list(L,_)).
```
```
?- test. 
15 inferences, 0.000 CPU in 0.000 seconds (113% CPU, 660153 Lips)
```
without
```
2,001 inferences, 0.000 CPU in 0.000 seconds (101% CPU, 4730183 Lips)
```
results are almost the same in case the list does not contains 0 elements (list of 10 000 elements generated where each element is random between 1 and 3).
```
20,001 inferences, 0.009 CPU in 0.009 seconds (100% CPU, 2214337 Lips) (with)
20,001 inferences, 0.009 CPU in 0.009 seconds (100% CPU, 2349666 Lips) (without)
```